File: src/main/java/edu/uchicago/gerber/_08final/mvc/controller/Sound.java
package edu.uchicago.gerber._08final.mvc.controller;


import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Executors;
import java.util.concurrent.ThreadPoolExecutor;
import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.Clip;
import javax.sound.sampled.LineUnavailableException;
import javax.sound.sampled.UnsupportedAudioFileException;

public class Sound {

	//for sound playing. Limit the number of threads to 5 at a time.
	private static final ThreadPoolExecutor soundExecutor = (ThreadPoolExecutor) Executors.newFixedThreadPool(5);

	//A Looped thread is one that plays for an indefinite time until
	// you call the .stopSound() method.
	public static final Map<String, Clip> LOOP_SOUNDS;

	// Load all looping sounds in the static context.
	static {
		Path rootDirectory = Paths.get("src/main/resources/sounds");
		Map<String, Clip> localMap = null;
		try {
			localMap = loadLoopedSounds(rootDirectory);
		} catch (IOException e) {
			e.fillInStackTrace();
			throw new ExceptionInInitializerError(e);
		}
		LOOP_SOUNDS = localMap;

	}

	private static Map<String, Clip> loadLoopedSounds(Path rootDirectory) throws IOException {
		Map<String, Clip> soundClips = new HashMap<>();
		Files.walkFileTree(rootDirectory, new SimpleFileVisitor<Path>() {
			@Override
			public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
				if (file.toString().toLowerCase().endsWith("_loop.wav")) {
					try {
						Clip clip = getLoopClip(file);
						if (clip != null) {
							soundClips.put(file.getFileName().toString(), clip);
						}
					} catch (Exception e) {
						e.fillInStackTrace();
					}
				}
				return FileVisitResult.CONTINUE;
			}

			@Override
			public FileVisitResult visitFileFailed(Path file, IOException exc) {
				exc.fillInStackTrace();
				return FileVisitResult.CONTINUE;
			}
		});
		return soundClips;
	}

	private static Clip getLoopClip(Path fileName) throws Exception {
		Clip clip = null;
		try {
			// Adjust the path to be relative to the resources directory
			String relativePath = "/sounds/" + fileName.getFileName().toString();
			InputStream audioSrc = Sound.class.getResourceAsStream(relativePath);

			if (audioSrc == null) {
				throw new IOException("No such sound file exists at " + relativePath);
			}

			InputStream bufferedIn = new BufferedInputStream(audioSrc);
			AudioInputStream aisStream = AudioSystem.getAudioInputStream(bufferedIn);
			clip = AudioSystem.getClip();
			clip.open(aisStream);

		} catch (UnsupportedAudioFileException | IOException | LineUnavailableException e) {
			e.fillInStackTrace();
			throw e;
		}

		return clip;
	}

	public static void stopSound(final String strPath) {
		//non-looped sounds can not be 'stopped', they simply expire.
		if (strPath.contains("_loop")) {
			LOOP_SOUNDS.get(strPath).stop();
		}
	}


	//Sounds that can be played simultaneously, must be queued onto the soundExecutor at runtime.
	public static void playSound(final String strPath) {
		if (strPath.contains("_loop")){
			LOOP_SOUNDS.get(strPath).loop(Clip.LOOP_CONTINUOUSLY);
			return;
		}

		soundExecutor.execute(new Runnable() {
			public void run() {
				try {
					Clip clp = AudioSystem.getClip();

					InputStream audioSrc = Sound.class.getResourceAsStream("/sounds/" + strPath);
                    assert audioSrc != null;
                    InputStream bufferedIn = new BufferedInputStream(audioSrc);
					AudioInputStream aisStream = AudioSystem.getAudioInputStream(bufferedIn);

					clp.open(aisStream);
					clp.start();
				} catch (Exception e) {
					System.err.println(e.getMessage());
				}
			}
		});

	}



}

File: src/main/java/edu/uchicago/gerber/_08final/mvc/controller/Sound.kt
package edu.uchicago.gerber._08final.mvc.controller

import lombok.Synchronized
import java.io.BufferedInputStream
import java.io.IOException
import java.io.InputStream
import javax.sound.sampled.AudioSystem
import javax.sound.sampled.Clip
import javax.sound.sampled.LineUnavailableException
import javax.sound.sampled.UnsupportedAudioFileException

object Sound {
    //for individual wav sounds (not looped)
    //http://stackoverflow.com/questions/26305/how-can-i-play-sound-in-java
    fun playSound(strPath: String) {
        //use coroutines here on the io dispatcher
        Thread {
            try {
                val clp = AudioSystem.getClip()
                val audioSrc = Sound::class.java.getResourceAsStream("/sounds/$strPath")
                val bufferedIn: InputStream = BufferedInputStream(audioSrc)
                val aisStream = AudioSystem.getAudioInputStream(bufferedIn)
                clp.open(aisStream)
                clp.start()
            } catch (e: Exception) {
                System.err.println(e.message)
            }
        }.start()
    }

    //for looping wav clips
    //http://stackoverflow.com/questions/4875080/music-loop-in-java
    fun clipForLoopFactory(strPath: String): Clip {
        var clp: Clip? = null
        try {
            val audioSrc = Sound::class.java.getResourceAsStream("/sounds/$strPath")

            val bufferedIn: InputStream = BufferedInputStream(audioSrc)
            val aisStream = AudioSystem.getAudioInputStream(bufferedIn)
            clp = AudioSystem.getClip()
            clp.open(aisStream)
        } catch (e: UnsupportedAudioFileException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: LineUnavailableException) {
            e.printStackTrace()
        }
        return clp!!
    }
}
